File: rtmb_java/src/com/ibm/realtime/rtmb/tests/TimerConfig.java
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
package com.ibm.realtime.rtmb.tests;

import com.ibm.realtime.rtmb.params.RealTimeValues;
import com.ibm.realtime.rtmb.params.TestInfo;
import com.ibm.realtime.rtmb.params.TestParams;
import com.ibm.realtime.rtmb.results.ResultTable;
import com.ibm.realtime.rtmb.results.TestResults;
import com.ibm.realtime.rtmb.tests.util.SpecTest;
import com.ibm.realtime.rtmb.tests.util.TestFinalDefinitions;
import com.ibm.realtime.rtmb.tests.util.TestId;
import com.ibm.realtime.rtmb.util.VerbosePrint;

public class TimerConfig extends SpecTest {
	private double max(double val1, double val2) {
		return (val1 > val2) ? val1 : val2;
	}

	private double min(double val1, double val2) {
		return (val1 < val2) ? val1 : val2;
	}

	public boolean runTimerTest(int num, TestResults result,
			TestParams tParams, int new_iters) {
		long startTime, endTime;
		long timeDiff = 0;
		int loop = 0;
		
		for (loop = 0; loop < new_iters; loop++) {
			startTime = System.nanoTime();
			try {
				Thread.sleep(0, num); // do some nano sec sleep
			} catch (InterruptedException e) {
				e.printStackTrace();
				return false;
			}
			endTime = System.nanoTime();

			/* Get the elapsed time in nsecs */
			timeDiff = endTime - startTime;
			//msf - a little too noisy - VerbosePrint.diffTime(System.out, tParams, name(), timeDiff);
			
			result.setValuesAtNextIndex(timeDiff);
		}

		return true;
	}

	@Override
	public TestResults[] benchMarkTest(TestInfo tInfo, RealTimeValues rtValues,
			TestParams tParams) {
		int count = 0;

		if (timerAccuracySpecified()) {
			VerbosePrint.streamPrint (System.out, tParams, 1, 
			"TimerConfig: Configuration test will not be run. Timer accuracy preconfigured as: ");
			VerbosePrint.streamPrintf(System.out, tParams, 1, "%.3f us" + 
					VerbosePrint.getLineSeparator(), 
					VerbosePrint.roundToMicroSec(getTimerAccuracy()));
			VerbosePrint.outPrintln (System.out, tParams, 1);
			
			return null;
		}
		

		
		double clock = ClockConfig.getClockAccuracy();
		int num = TestFinalDefinitions.TIMER_CONFIG_NSLEEP_TIME;
		int new_iters;

		boolean retryMsg = false;

		new_iters = updateIterations(tInfo, rtValues, tParams);
		TestResults[] testResult = establishSingleResult(TestId.SYSTIMER_CONFIG, tInfo.getTestIterations());

		VerbosePrint.testStart(System.out, tParams, name(), tInfo.getTestIterations());
		
		double savedAcceptablePercent = rtValues.getAcceptablePercent();
		double savedRangePercent = rtValues.getRangePercent();
		boolean flag = false;
		if (savedAcceptablePercent < 99.9 || savedRangePercent > 20.0) {
			flag = true;
		}
		
		double maximumPct = max(99.9, rtValues.getAcceptablePercent());
		rtValues.setAcceptablePercent(maximumPct);
		
		double minPct = min(20.0, rtValues.getRangePercent());
		rtValues.setRangePercent(minPct);
		if (flag) {
			VerbosePrint.outPrintln(System.out, tParams, 1);
			VerbosePrint.streamPrintln(System.out, tParams, 1,
					"TimerConfig: Altered the acceptable percentage to: "
							+ rtValues.getAcceptablePercent());
			VerbosePrint.streamPrintln(System.out, tParams, 1,
					"TimerConfig: Altered the range percentage to: "
							+ rtValues.getRangePercent());
			VerbosePrint.outPrintln(System.out, tParams, 1);
		}

		boolean testPass = false;
		int previousQuantum = 0;

		while (true) {
			boolean rc = false;
			if (count++ < tInfo.getThreshold()) {
				testResult[0].resetResults();

				rc = runTimerTest(num, testResult[0], tParams, new_iters);
				if (rc == false) {
					return null;
				}

				if (testResult[0].getMedian() > clock) {
					retryMsg = true;

					/* Check against the input parameters for this test */
					if (testResult[0].checkStdDev(rtValues, tParams)) {
						/* Variance looks good */
						retryMsg = false;
						testPass = true;
						break;
					}
				}
			} else {
				break;
			}

			if (retryMsg) {
				VerbosePrint.testRetry(System.out, tParams, name());
			}

			num = num * 2;
		}

		testResult[0].setWorkQuantum(num);

		if (testPass) {
			TestResults passedResult = null;
			int startQuantum = previousQuantum;
			int stopQuantum = num;
			int midQuantum = 0;

			passedResult = new TestResults(TestId.SYSTIMER_CONFIG, tInfo
					.getTestIterations(), true);
			passedResult.cloneValues(testResult[0]);

			TestResults tResult = new TestResults(TestId.SYSTIMER_CONFIG, tInfo
					.getTestIterations(), true);

			while (startQuantum <= stopQuantum) {
				tResult.resetResults();
				midQuantum = (startQuantum + stopQuantum) / 2;

				VerbosePrint.streamPrintln(System.out, tParams, 1,
						"TimerConfig: Refine Work Quantum. Range from: " + startQuantum + " to: " + stopQuantum);

				runTimerTest(midQuantum, tResult, tParams, new_iters);

				if (tResult.checkStdDev(rtValues, tParams)) {
					passedResult.resetResults();
					tResult.setWorkQuantum(midQuantum);
					passedResult.cloneValues(tResult);
					stopQuantum = midQuantum - 1;
				} else {
					startQuantum = midQuantum + 1;
				}
			}

			testResult[0] = passedResult;
		}

		VerbosePrint.streamPrintln(System.out, tParams, 1,
				"TimerConfig: WorkQuantum = " + testResult[0].getWorkQuantum());

		VerbosePrint.streamPrint(System.out, tParams, 1,
				"TimerConfig: Benchmark will be configured "
						+ "to expect the timer accuracy to be accurate to ");
		VerbosePrint.streamPrintf(System.out, tParams, 1, "%.3f us"
				+ VerbosePrint.getLineSeparator(), 
				VerbosePrint.roundToMicroSec(testResult[0].getMedian()));
		VerbosePrint.outPrintln(System.out, tParams, 1);

		if (flag) {
			rtValues.setAcceptablePercent(savedAcceptablePercent);
			rtValues.setRangePercent(savedRangePercent);
		}

		return testResult;
	}

	/**
	 * 
	 * @return The accuracy of the timer, which is currently given by the median
	 *         result returned from running the timer configuration test. 
	 */
	private static double accuracy = 0.0;
	public static double getTimerAccuracy() {
		if (accuracy == 0.0) {
			accuracy = (ResultTable.getInstance().getTestResultsByTestID(TestId.SYSTIMER_CONFIG)[0].getMedian());
			System.out.println("TimerConfig Accuracy: " + accuracy);
		}
		return accuracy;
	}

	/**
	 * getWorkQuantum - msf - need something less vague here
	 * @return a work quantum, based on the timer accuracy, that is large enough to represent reasonable work
	 */
	public static double getWorkQuantum() {
		return getTimerAccuracy();
		/*
		if (quantum == 0.0) {
			quantum = (ResultTable.getInstance().getTestResultsByTestID(TestId.SYSTIMER_CONFIG)[0].getWorkQuantum());
			System.out.println("TimerConfig Accuracy: " + getTimerAccuracy());
			System.out.println("TimerConfig Quantum: " + quantum);
		}
		return quantum;
		*/
	}	

	/**
	 * Explicitly set the timer accuracy from a command-line option
	 * 
	 * @param time : String representing the timer accuracy in the form <xx>[ms|us|ns], e.g. 10us
	 * @return : true if the timer was not set (because the format is invalid), false otherwise
	 */
	public static boolean setTimerAccuracy(String time) {
		long multiplier;
		if (time.endsWith("ns")) {
			multiplier = 1L;
		} else if (time.endsWith("us")) {
			multiplier = 1000L;
		} else if (time.endsWith("ms")) {
			multiplier = 1000000L;
		} else {
			return true;
		}
		long value = Long.parseLong(time.substring(0,time.length()-2).trim());
		if (value <= 0L) {
			return true;
		}
		accuracy = (double) (value * multiplier);
		return false;
	}
	
	public static boolean timerAccuracySpecified() {
		return accuracy != 0.0;
	}
	
	
	@Override
	public String name() {
		return "TimerConfig";
	}
}

File: rtmb_c/src/testcase/config_test/timer_config.c
/*******************************************************************************
 * Copyright (c) 2008 International Business Machines Corp.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Ajith Ramanath            - initial API and implementation
 *    Radhakrishnan Thangamuthu - initial API and implementation
 *    Mike Fulton               - initial API and implementation
 *******************************************************************************/
 
#include <config_test.h>

void run_timer_test(int num, int iters, test_results_t *result,
        cmd_line_args *cmd_args, benchmark_test* bmtest) {
	int loop;

	struct timespec req_time, start_time, end_time;
	long long time_diff;

	req_time.tv_nsec = num % 1000000000LL;
	req_time.tv_sec = num / 1000000000LL;

	for (loop = 0; loop < iters; loop++) {
		if (get_cur_time(&start_time) == ERROR) {
			abort();
		}

		if (nanosleep(&req_time, NULL) != 0) {
			RTMB_printf(stderr, " nanosleep failed.. Exiting \n");
			abort();
		}

		if (get_cur_time(&end_time) == ERROR) {
			abort();
		}

		/* Get the time for which we actually slept in nsecs */
		time_diff = get_time_diff(start_time, end_time);
		RTMB_verbose_printf(stdout, cmd_args, 2,
		        "timer_config: Difference between end and start"
			        " times = %.3f us \n", MICROSEC(time_diff));

		/*
		 * We would have definitely slept slightly more than
		 * NANOSLEEP_TIME.
		 */
		add_entry(result, time_diff, 0);
	}
}
/*
 * Name    : timer_config()
 * Desc    : Determines accuracy of the timer.
 * Args    : args - Arguments to the test.
 * Returns : NULL
 */
void *timer_config(void *args) {
	/* Get the args passed from parent thread */
	thread_args_t *targs = (thread_args_t *) args;
	cmd_line_args *cmd_args = targs->cmd_args;
	benchmark_test *bmtest = targs->bmtest;
	result_table_t *res_table = targs->res_table;
	test_results_t *result = NULL;
	double clock_accuracy;
	int start_quantum, stop_quantum, mid_quantum;
	test_results_t *passed_result;
	char print_flag = FALSE;

	/* Other variables local to this function */
	int done = 0, count = 0;
	int num = NANOSLEEP_TIME;
	int prev_num = 0;
	int new_iters = bmtest->_iterations;
	double saved_percentage;
	double saved_range;
	double change_percent = 0;
	int test_pass;

	/*
	 * fill the rt structures with correct percentage
	 * and range for config tests.
	 * Since the same rt structure is used by all tests
	 * do not forget to retain the old values for percentages
	 * and values when the code returns from this function
	 */

	if (get_acceptable_percent(bmtest, cmd_args) < ACCEPTABLE_PCT_CFG_TEST
	        || get_range_percent(bmtest, cmd_args)
	                > ACCEPTABLE_RANGE_CFG_TEST) {
		change_percent = 1;
	} else {
		change_percent = 0;
	}

	if (change_percent) {
		fill_acceptable_percentages(bmtest, cmd_args,
		        ACCEPTABLE_PCT_CFG_TEST, ACCEPTABLE_RANGE_CFG_TEST,
		        &saved_percentage, &saved_range);
	}

	result = create_test_result(1, new_iters);
	passed_result = create_test_result(1, new_iters);
	if (result == NULL || passed_result == NULL) {
		RTMB_printf(stderr,
		        "ERROR: Cannot allocate memory for test_results_t");
		RTMB_printf(stderr, " in timer_config\n");
		abort();
	}

	strcpy((char *) result->desc, (char *) "Timer configuration test");
	RTMB_verbose_printf(stdout, cmd_args, 1, "\nTest Report for %s:\n",
	        (char*) &result->desc);
	RTMB_verbose_printf(stdout, cmd_args, 1,
	        "=========================================\n");
	RTMB_verbose_printf(stdout, cmd_args, 1,
	        "\ntimer_config : Total number of iterations = %d\n\n",
	        new_iters);

	clock_accuracy = get_min_exec_time(res_table);

	while (!done) {
		RTMB_verbose_printf(stdout, cmd_args, 3,
		        "timer_config: Requested sleep() duration: %d ns\n",
		        num);

		run_timer_test(num, new_iters, result, cmd_args, bmtest);

		if (IS_EXEC_TIME_GRT_THAN_CLK(result, clock_accuracy)) {
			print_flag = TRUE;
			if (check_pass_criteria(result, cmd_args, bmtest, 0)
			        == SUCCESS) {
				/*
				 * Std deviation looks good,
				 * accuracy of timer is determined
				 */
				test_pass = ACCEPTABLE_DEVIATIONS;
				done = 1;
				break;
			} else {
				count++;
				/* Return error if we have tried enough */
				if (count == bmtest->_threshold) {
					RTMB_printf(stderr,
					        "timer_config: exceeded "
						        "maximum attempts \n");
					test_pass = UNACCEPTABLE_DEVIATIONS;
					done = 1;
					break;
				}

			}
		}

		/*
		 * Std dev is not accurate enough,
		 * so retry with larger sleep time
		 */
		free_chain(result, 0);
		prev_num = num;
		num = num * MULTIPLIER_FOR_SUB_ITER;

		if (print_flag) {
			RTMB_verbose_printf(stdout, cmd_args, 1,
			        "\ntimer_config: Retrying test ");
			RTMB_verbose_printf(stdout, cmd_args, 1,
			        "with bigger work quantum to get "
				        "lesser variance...\n");
		}
	}

	result->opern_amount = num;

	if (test_pass == ACCEPTABLE_DEVIATIONS) {
		/*
		 * Perform binary search and find out least quantum of work
		 * that needs to be done in order to get the test pass.
		 */
		start_quantum = prev_num;
		stop_quantum = num;

		clone_results(passed_result, result);

		/*
		 * Run while loop
		 * for all possible quantums
		 */
		while (start_quantum <= stop_quantum) {
			char binary_search_test_pass = FALSE;
			free_chain(result, 0);
			mid_quantum = (start_quantum + stop_quantum) / 2;

			RTMB_verbose_printf(
			        stdout,
			        cmd_args,
			        3,
			        " start_quantum = %d mid_quantum = %d stop_quantum = %d \n",
			        start_quantum, mid_quantum, stop_quantum);

			run_timer_test(mid_quantum, new_iters, result,
			        cmd_args, bmtest);

			if (IS_EXEC_TIME_GRT_THAN_CLK(result, clock_accuracy)
			        && (check_pass_criteria(result, cmd_args,
			                bmtest, 0) == SUCCESS)) {
				binary_search_test_pass = ACCEPTABLE_DEVIATIONS;
			} else {
				binary_search_test_pass
				        = UNACCEPTABLE_DEVIATIONS;
			}

			if (binary_search_test_pass == UNACCEPTABLE_DEVIATIONS) {
				start_quantum = mid_quantum + 1;
			} else if (binary_search_test_pass
			        == ACCEPTABLE_DEVIATIONS) {
				result -> opern_amount = mid_quantum;
				clone_results(passed_result, result);
				stop_quantum = mid_quantum - 1;
			}
		}

		compute_std_time(passed_result, 0);

		delete_test_result(result);
		result = passed_result;
	}

	RTMB_verbose_printf(stdout, cmd_args, 1,
	        "timer_config: Benchmark will be configured ");
	RTMB_verbose_printf(stdout, cmd_args, 1,
	        "to expect the timer to be accurate to ");
	RTMB_verbose_printf(stdout, cmd_args, 1, "%.3f us\n",
		MICROSEC(result->mediantime));

	add_result_to_result_table2(res_table, result, SYSTIMER_CONFIG, bmtest);

	if (change_percent) {
		/* revert the percentages and ranges to original values */
		fill_acceptable_percentages(bmtest, cmd_args, saved_percentage,
		        saved_range, NULL, NULL);
	}

	fflush(stderr);
	fflush(stdout);
	targs->ret = SUCCESS;
	return (void *) NULL;
}
